chaste_codegen/be_model.py

```diff
@@ -0,0 +1,314 @@
+import chaste_codegen as cg
+import time
+import sympy as sp
+from chaste_codegen._partial_eval import partial_eval
+from enum import Enum
+from sympy.codegen.cfunctions import (log10, log2)
+
+
+class BeModel(cg.ChasteModel):
+    """ Holds template and information specific for the Backwards Euler model type"""
+
+    def __init__(self, model, file_name, **kwargs):
+        super().__init__(model, file_name, use_analytic_jacobian=True, **kwargs)
+        # get deriv eqs and substitute in all variables other than state vars
+        self._derivative_equations = \
+            partial_eval(self._derivative_equations, self._y_derivatives, keep_multiple_usages=False)
+        self._non_linear_state_vars = self._get_non_linear_state_vars()
+        self._jacobian_equations, self._jacobian_matrix = self._get_jacobian()
+        self._formatted_state_vars, self._formatted_nonlinear_state_vars, self._formatted_residual_equations, \
+            self._formatted_derivative_eqs = self._update_state_vars()
+        self._formatted_rearranged_linear_derivs, self._formatted_linear_equations = \
+            self._format_rearranged_linear_derivs()
+        self._formatted_jacobian_equations, self._formatted_jacobian_matrix_entries = self._format_jacobian()
+
+    class KINDS(Enum):
+        none = 1
```

Comment:
You might want to call these `NONE`, `LINEAR`, `NONLINEAR` so the capitalisation is consistent.